scrapy_po/utils.py

```diff
@@ -1,7 +1,131 @@
 # -*- coding: utf-8 -*-
+import inspect
+from typing import Tuple, Dict, Type, Callable
+
+import andi
+from scrapy.http import Response
+from scrapy.utils.defer import maybeDeferred_coro
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from scrapy_po.webpage import Injectable
+from scrapy_po.page_input_providers import providers
+
 
 def get_callback(request, spider):
-    """ Get request.callback of a scrapy.Request, as a callable """
+    """Get request.callback of a scrapy.Request, as a callable."""
     if request.callback is None:
         return getattr(spider, 'parse')
     return request.callback
+
+
+class DummyResponse(Response):
+
+    def __init__(self, url, request=None):
+        super(DummyResponse, self).__init__(url=url, request=request)
+
+
+def is_callback_using_response(callback: Callable):
+    """Check whether the request's callback method is going to use response."""
+    spec = inspect.getfullargspec(callback)
+    try:
+        arg_name = spec.args[1]  # first index is self, second is response
+    except IndexError:
+        # Parse method is probably using *args and **kwargs annotation.
+        # Let's assume response is going to be used.
+        return True
+
+    if arg_name not in spec.annotations:
+        # There's no type annotation, so we're probably using response here.
+        return True
+
+    if issubclass(spec.annotations[arg_name], DummyResponse):
+        # Type annotation is DummyResponse, so we're probably NOT using it.
+        return False
+
+    # Type annotation is not DummyResponse, so we're probably using it.
+    return True
+
+
+def is_provider_using_response(provider):
+    """Check whether injectable provider makes use of a valid Response."""
+    for argument, possible_types in andi.inspect(provider.__init__).items():
+        for cls in possible_types:
```

Comment:
I think we should fail with a useful message here in case `cls` is not expected (i.e. it is not any type of Response).

Reply:
Are we allowing other types of injectables in the future? 

Perhaps it's out of the scope of this method having to check this. What do you think? Should we raise a TypeError here?

Reply:
I'm thinking in the case where somebody declares a provider with a wrong `__init__`. It would be more meaningful to fail here with a TypeError saying something like: `Type foo not supported for arguments for providers __init__. Only Response is supported.` or `Too many arguments for provider __init__, only empty or one supported`. 

Otherwise, it will fail on invocation but the message will be much less informative. What do you think?

Reply:
Well, it makes sense. But I still think it's out of the scope of this method.

Think of that Config injectable you've suggested in another comment. Does that make sense to check for sub-classes of this Config here too? I don't think so. 

But let me know if you have a different opinion. Also, your two cents here would be helpful, @kmike.

Reply:
Ok, forget about it. I think I'm messing things here :stuck_out_tongue_closed_eyes:  and this is very minor.